### PR TITLE
Replace assertions with ValueErros on distilbert model

### PR DESCRIPTION
**Pull request by JuheonChu**
Co-author: @batese2001
This is an extended PR from [here](https://github.com/huggingface/transformers/pull/20432)  for new valid-checks to improve _code quality._

This raises exceptions on Multi-head Attention models for following specified conditions on the _Distilbert_ model.

This successfully replaces assertions with ValueErrors with the _Distilbert_ model related to #12789.

Would you be open to me changing assertions if I encounter other ones?

To: @younesbelkada

### CODE REVIEW

**Comment by JuheonChu** [comment]
@[Sylvain Gugger](https://github.com/sgugger) 
This is a new PR of my own based on your suggestion from [here](https://github.com/huggingface/transformers/pull/20375). files which improved to pass all the validity checks. 
Can I ask for your further suggestions for this PR to be merged? 
Any comments will be very appreciated!

**Comment by younesbelkada** [comment]
Hi @JuheonChu ! 👋
Thanks for the PR ;) will review it asap and let you know!

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 1]
```diff
@@ -141,7 +142,11 @@ def __init__(self, config: PretrainedConfig):
         self.dim = config.dim
         self.dropout = nn.Dropout(p=config.attention_dropout)
 
-        assert self.dim % self.n_heads == 0
+        # Have an even number of multi heads that divide the dimensions
+        if self.dim % self.n_heads != 0:
+            raise ValueError(
+                "self.n_heads (" + str(self.n_heads) + ") must divide self.dim (" + str(self.dim) + ") evenly"
```
Comment (younesbelkada) [inline src/transformers/models/distilbert/modeling_distilbert.py:148]:
```suggestion
                f"self.n_heads: {self.n_heads} must divide self.dim: {str(self.dim)} evenly"
```
It is preferable to use f-strings ;) check for example [here](https://github.com/huggingface/transformers/blob/3c39c07f1160beb6190f635e2dab6a27e6d67596/src/transformers/models/opt/modeling_opt.py#L139)

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 2]
```diff
@@ -255,7 +260,11 @@ class TransformerBlock(nn.Module):
     def __init__(self, config: PretrainedConfig):
         super().__init__()
 
-        assert config.dim % config.n_heads == 0
+        # Have an even number of Configure multi-heads
+        if config.dim % config.n_heads != 0:
+            raise ValueError(
+                "config.n_heads (" + str(config.n_heads) + ") must divide config.dim (" + str(config.dim) + ") evenly"
```
Comment (younesbelkada) [inline src/transformers/models/distilbert/modeling_distilbert.py:266]:
```suggestion
                f"config.n_heads {config.n_heads} must divide config.dim {config.dim} evenly"
```
same as above!

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 3]
```diff
@@ -291,7 +300,9 @@ def forward(
         if output_attentions:
             sa_output, sa_weights = sa_output  # (bs, seq_length, dim), (bs, n_heads, seq_length, seq_length)
         else:  # To handle these `output_attentions` or `output_hidden_states` cases returning tuples
-            assert type(sa_output) == tuple
+            if type(sa_output) != tuple:
+                raise TypeError("sa_output must be a tuple but it is " + str(type(sa_output)) + " type")
```
Comment (younesbelkada) [inline src/transformers/models/distilbert/modeling_distilbert.py:304]:
```suggestion
                raise TypeError(f"sa_output must be a tuple but it is {type(sa_output)} type")
```
same as above!

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 4]
```diff
@@ -348,11 +360,18 @@ def forward(
             hidden_state = layer_outputs[-1]
 
             if output_attentions:
-                assert len(layer_outputs) == 2
+                if len(layer_outputs) != 2:
+                    raise ValueError(
+                        "The length of the layer_outputs should be 2, but it is " + str(len(layer_outputs))
```
Comment (younesbelkada) [inline src/transformers/models/distilbert/modeling_distilbert.py:365]:
```suggestion
                       f"The length of the layer_outputs should be 2, but it is {len(layer_outputs)}"
```

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 5]
```diff
@@ -348,11 +360,18 @@ def forward(
...
+                    )
+
                 attentions = layer_outputs[0]
                 all_attentions = all_attentions + (attentions,)
             else:
-                assert len(layer_outputs) == 1
+                if len(layer_outputs) != 1:
+                    raise ValueError(
+                        "The length of the layer_outputs should be 1, but it is " + str(len(layer_outputs))
```
Comment (younesbelkada) [inline src/transformers/models/distilbert/modeling_distilbert.py:373]:
```suggestion
                        "The length of the layer_outputs should be 1, but it is {len(layer_outputs)}"
```

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 6]
```diff
@@ -810,7 +829,9 @@ def __init__(self, config: PretrainedConfig):
 
         self.distilbert = DistilBertModel(config)
         self.qa_outputs = nn.Linear(config.dim, config.num_labels)
-        assert config.num_labels == 2
+        if config.num_labels != 2:
+            raise ValueError("config.num_labels should be 2, but it is " + str(config.num_labels))
```
Comment (younesbelkada) [inline src/transformers/models/distilbert/modeling_distilbert.py:833]:
```suggestion
            raise ValueError("config.num_labels should be 2, but it is {config.num_labels)}"
```

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for working on this! 💪 
Left a couple of comments ;) we prefer to use f-strings instead, you can check for instance [what is done in OPT](https://github.com/huggingface/transformers/blob/main/src/transformers/models/opt/modeling_opt.py#L139-L141)

**Comment by JuheonChu** [comment]
Thank you @younesbelkada  for providing us with valuable suggestions! We will make changes and make anotehr PR! 
Sincerely appreciate it :)

**Comment by younesbelkada** [comment]
No worries @JuheonChu ! 
As I made you some suggestions, actually you can directly continue on this PR, let me guide you through this step by step: 
Step 1: Go to "file changed" (top right of the github UI of this PR):
Screenshot 2022-11-25 at 22 17 47

Step 2: For each suggestion, click on "Add suggestion to batch", to add each suggestion
Screenshot 2022-11-25 at 22 18 59

Step 3: Once you have added all the suggestions (make sure to add all of them), a pop up bar will appear on the top-right corner, and you can just click to it, and the suggestions will be pushed ;) 

Screenshot 2022-11-25 at 22 19 43

This way no need to open a new PR each time we do a suggestion ;) ! Let me know if anything is unclear!

File: src/transformers/models/distilbert/modeling_distilbert.py (hunk 6 above)
Comment (JuheonChu) [reply src/transformers/models/distilbert/modeling_distilbert.py:833]: I think you missed the opening parenthesis, but I sincerely appreciate this suggestion!

Comment (younesbelkada) [reply src/transformers/models/distilbert/modeling_distilbert.py:833]: Ah yes you are right I missed the parenthesis here ! :)

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 7]
```diff
@@ -810,7 +829,9 @@ def __init__(self, config: PretrainedConfig):
 
         self.distilbert = DistilBertModel(config)
         self.qa_outputs = nn.Linear(config.dim, config.num_labels)
-        assert config.num_labels == 2
+        if config.num_labels != 2:
+            raise ValueError(f"config.num_labels should be 2, but it is {config.num_labels)}"
```
Comment (JuheonChu) [inline src/transformers/models/distilbert/modeling_distilbert.py:833]:
```suggestion
            raise ValueError ( f"config.num_labels should be 2, but it is {(config.num_labels)}")

**Review by JuheonChu — COMMENTED** [review]
If we change Line 833 with raise ValueError ( f"config.num_labels should be 2, but it is {(config.num_labels)}"), I think the checks will all pass!

**Comment by JuheonChu** [comment]
Due to the reformatting of jupyter files, I created a new PR [here](https://github.com/huggingface/transformers/pull/20463).
Now it passes all the validation checks. I apologize for any inconvenience, but do you mind if you can check  [here](https://github.com/huggingface/transformers/pull/20463)?

To: @younesbelkada

**Comment by younesbelkada** [comment]
Sure! I propose to close this PR as we moved everything on the other PR ;)

**Comment by JuheonChu** [comment]
Closing this PR. This PR is solved in #20463.

**Pull request closed without merging**